### PR TITLE
Added the TextureStitchEvent.Collected event.

### PR DESCRIPTION
**Pull request by marchermans**
When @liach moved the TextureStitchEvent.Pre call to before the collection of the TextureAtlasSprites of Items and Blocks, in PR #3436, he made it also impossible to register additional Textures that are based on already registered TextureAtlasSprites, as the registered list in the Map is now empty, and there is no way to retrieve an already registered Texture as none have bene registered.

A good example are the materialized greyscale textures of TinkersConstruct, currently maintained by @bonii-xx, or my mod called Armory.

This PR creates a new TextureStitchEvent.Collected that is mend to be fired after the TextureAtlasSprites of the Blocks and Items have been collected the normal way, but before the Stitch actually happens. This makes it possible to register textures based on already registered Textures, before the Stitch happens.

I added a method to ForgeHooksClient to minimalize the changes to the patch in TextureMap,java and added the call to that method on the location of old call to ForgeHooksClient.onTextureStitchPre.

_**UPDATE:**_
It has been pointed out to me that I should explain my reason why i wanted this PR better, as for a simple recolor of an existing texture I could use IItemColor. I agree that this would be the better solution if i just wanted to recolor existing textures with an overlay color. However this is not the only use case for my system that requires this event. I have, next to metalic materials for whom IItemColor whould have been enough, also materials that require an existing texture as the overlay. A good example is Wood. The bark texture of the material is then used as the overlay and retextures the existing greyscale model pixel for pixel. This is not possible by simply using IItemColor as that only allows for individual colors.

_**EXAMPLE:**_
An example of a TextureCreator that is part of my Library SmithsCore: https://github.com/SmithsGaming/SmithsCore/blob/Development-1.11/src/com/smithsmodding/smithscore/client/textures/TextureCreator.java is used to create the individual textures. In this Example a modder can register an Implementation of ITextureController to modify marked textures and reregister them.

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java [hunk 1]
```diff
@@ -48,6 +48,16 @@ public TextureMap getMap()
     }
 
     /**
+     * Fired when the TextureMap is told to refresh it's stitched texture.
+     * Called after all textures for the map have been collected, but before the stitching
+     * is done.
+     */
+    public static class Collected extends TextureStitchEvent
+    {
+        public Collected(TextureMap map){ super(map); }
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java]: Code style

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
> A good example are the materialized greyscale textures of TinkersConstruct, currently maintained by @bonii-xx, or my mod called Armory.

Can you give a link to the code?

File: src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java (hunk 1 above)
Comment (marchermans) [reply src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java]: What is wrong here? The formatting is the same in the other Events.

**Comment by marchermans** [comment]
Added an example and better reasoning why i need the Event.

File: src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java (hunk 1 above)
Comment (liach) [reply src/main/java/net/minecraftforge/client/event/TextureStitchEvent.java]: Well, yes, all the events are improperly formatted. You can ignore this point, indeed.

**Comment by liach** [comment]
I still wonder how you use these sprites without references in jsons. That was why I moved this event.

**Comment by marchermans** [comment]
The principal system here works in combination with an API and a custom ModelLoader.
The API holds reference to all possible combination that can be made on the Armor in other words:
 - Which materials can be made into the given Base Armor.
 - Which upgrades can be put onto this armor and out of which Materials are they made.

The ModelLoader then loads a JSON which reflects these informations. It holds a base texture for the base Armor, in greyscale, and textures for each upgrade, also in greyscale. (Using events during the loading of the model somebody who is registering additional upgrades for a model can also provide the greyscale textures for that upgrade during loading.) It then returns a unbaked model that has those greyscale textures as its textures.

Before the change those greyscale textures where registered to the Stitcher and I could grab them during the Pre event. Materialize them, so for example color them in a metal color or overlay them with an existing texture of say a wooden bark for a wooden material, and then register those materialized textures to the map so that they would be stitched in properly.

The fact that the baking of the models happens after the textures are loaded allows me to grab the materialized textures during bake and create a new copied model that is retextured with the correct materialized texture during the baking process and store them in a collected list. When handleItemState is called, using the data stored on the stack it will then grab first the correctly textured basemodels quad for the base armor, and then if applicable the quads of all the correctly materialized models of the upgrades and combines them to that it generates one flattened model.

Now that the TextureStitchEvent.Pre is moved to before the collection of the Textures from the Models. It is impossible to grab the greyscale textures that it needs to generate the materialized ones that depend on those. That is the reason why i added the collected event phase into the TextureStitch. At this point the greyscale textures are collected and i can grab them and then retexture those like i did with the Pre event before it was moved.

Greets, Orion

My implementation:
 * Loader: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/loaders/MultiLayeredArmorModelLoader.java
 * UnbakedModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/unbaked/MultiLayeredArmorItemModel.java
      * UnbakedComponentModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/unbaked/components/ArmorCoreComponentModel.java
 * BakedModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/baked/BakedMultiLayeredArmorItemModel.java
      * UnbakedComponentModel: https://github.com/SmithsGaming/Armory/blob/Development-1.10.2/src/main/com/smithsmodding/armory/client/model/item/baked/components/BakedCoreComponentModel.java

**Review by liach — APPROVED** [review]

**Review by Madsthunder — APPROVED** [review]

**Comment by marchermans** [comment]
Should I add a Test Mod for this?

**Comment by liach** [comment]
Absolutely!

**Comment by marchermans** [comment]
I added a test mod and pulled in the latest changes.
Examples of what the test mod does:

image

And:

image

**Comment by marchermans** [comment]
@liach @RainWarrior Should be good to go.

**Comment by liach** [comment]
looks great. Just waiting for fry then.

**Comment by Madsthunder** [comment]
@mezz Any idea when the next round of prs will be pulled?

**Comment by marchermans** [comment]
@mezz , is there any news on this?

**Comment by mezz** [comment]
Looks good but I think the test mod is not working as intended.
From the TextureDump output, I just see a square outline instead of the Book and Pickaxe outlines.
texture_atlas_dump_textures_mipmap_0

Also the test mod should use the same formatting as the rest of Forge, with braces on new lines.

**Comment by marchermans** [comment]
Hmm The pictures above were taken from the texture dump generated by the same test mod
Let me Check what is going on. I will update the formatting as well.

**Comment by marchermans** [comment]
I ran the tests again, and i am getting the same result as above.
I am trying to figure out what is happening.

**Comment by marchermans** [comment]
@mezz How are you running the test?

**Comment by liach** [comment]
Eh, the test look extremely complex to me.

**Comment by marchermans** [comment]
@liach it finds the edge of a texture by iterating over all its pixels.
It is not that complicated.

**Comment by mezz** [comment]
I use IntelliJ to run the test mods within the Forge dev environment.
Here's a video on how to set it up: https://www.youtube.com/watch?v=pLWQk6ed56Q

The trick is to build the Tests project so it creates all the classes, and then set the classes folder as a runtime dependency for the Main project. 



That way the test will be run when you run the main project, and to update them you just have to build the Test project.

**Comment by marchermans** [comment]
Ah okey
I will give that a try, i usually create a dummy copy in the main module.

**Comment by mezz** [comment]
I pulled the changes and it works now.
In the diff it looks like you replaced the two calls to `getAtlasSprite` with `getTextureExtry`, maybe that fixed it and you didn't commit the change before.

**Comment by mezz** [comment]
Tested and looks good.

**TLDR** for Lex (sorry Orion but you are extremely verbose!):
This event allows mods to create textures based on other textures in the atlas.
The example mod takes the book and pickaxe texture and creates a new texture that's just their outline.

**Comment by RainWarrior** [comment]
What happens if someone wants to depend on one of the textures added in this event? In the [example](https://github.com/SmithsGaming/SmithsCore/blob/Development-1.11/src/com/smithsmodding/smithscore/client/textures/TextureCreator.java#L93), you're using EventPriority.LOWEST, which, from my understanding, means that nothing else will ba able to run after you. Wouldn't a system with an explicit dependencies be better? An event solution like this doesn't actually solve the problem of texture dependencies, it postpones it. Unless I've missed something :)

**Comment by mezz** [comment]
What you linked to and are discussing is not the example used in this PR.

**Comment by marchermans** [comment]
@Mezz what do you mean?

**Comment by mezz** [comment]
Fry was looking at the old implementation and not the one in the PR. I talked to him on IRC and his argument still stands.

**Comment by marchermans** [comment]
@mezz @RainWarrior I changed it. I tested my test mod again and this works in both cases.
Also to display that event works for textures that depend on a other texture i added log output that shows you that the events gets fired properly for textures i register while i am in the event handler.

**Comment by tterrag1098** [comment]
@OrionDevelopment Thank you for this, the earlier PR breaks my json system as well, since I need to parse texture metadata for additional textures to stitch. So one more 👍 from me.

**Comment by Baylem** [comment]
@OrionDevelopment I love this, the previous PR by liach messed up my systems as well. so a :+1: from me

**Comment by liach** [comment]
To all of you @tterrag1098 @OhCreative, if I added one event back then (mezz actually suggested to mess this up, see comments) things would be fine!

**Comment by tterrag1098** [comment]
@liach I think these two PRs should have happened at once, there is good reason to have both phases, but clearly there are issues (which this PR solves) with texture dependencies.

**Comment by marchermans** [comment]
Sadly enough I am still waiting on it.

**Comment by marchermans** [comment]
@liach Things like this can happen, no way you could have for seen somebody actually using it to make texture dependencies. If this gets pulled then it is all forgiven and forgotten :D

**Comment by liach** [comment]
Waiting for fry. He seems to be away.

**Comment by mezz** [comment]
#4070 has been accepted and merged, any further improvements are welcome. 
I think it means that this PR can be closed, is that right?

**Comment by marchermans** [comment]
Yes i think so.
